### PR TITLE
Make spawned_money work with all models

### PR DESCRIPTION
**Pull request by ARitz-Cracker**
I was creating my own spawned money entity for ARCBank to make the thing cross-gamemode compatible. I decided to look at the DarkRP spawned money to see how that worked. So, I took that entity, added some stuff I did to my entity, and now I am giving back (part) of what I did to make it better!

### CODE REVIEW

File: entities/entities/spawned_money/cl_init.lua [hunk 1]
```diff
@@ -3,27 +3,25 @@ include("shared.lua")
 function ENT:Initialize()
 end
 
+local vector_up = Vector(0,0,1)
```
Comment (FPtje) [inline entities/entities/spawned_money/cl_init.lua]: - Not the same as it was before (`Pos + Ang:Up() * 0.9`)
- Doesn't make it more modifiable

File: entities/entities/spawned_money/cl_init.lua [hunk 2]
```diff
@@ -3,27 +3,25 @@ include("shared.lua")
...
 function ENT:Draw()
     self:DrawModel()
+    local offset = (self:OBBMaxs() - self:OBBCenter())*vector_up
```
Comment (FPtje) [inline entities/entities/spawned_money/cl_init.lua]: Spaces around operators. Install glualint.

File: entities/entities/spawned_money/cl_init.lua [hunk 3]
```diff
@@ -3,27 +3,25 @@ include("shared.lua")
...
-    local Pos = self:GetPos()
+    local Pos = self:OBBCenter() 
     local Ang = self:GetAngles()
 
     surface.SetFont("ChatFont")
     local text = DarkRP.formatMoney(self:Getamount())
     local TextWidth = surface.GetTextSize(text)
 
-    cam.Start3D2D(Pos + Ang:Up() * 0.9, Ang, 0.1)
+    cam.Start3D2D(self:LocalToWorld(Pos+offset), Ang, 0.1)
```
Comment (FPtje) [inline entities/entities/spawned_money/cl_init.lua]: Spaces around operators. Have you installed glualint yet?

File: entities/entities/spawned_money/cl_init.lua [hunk 4]
```diff
@@ -3,27 +3,25 @@ include("shared.lua")
...
         draw.WordBox(2, -TextWidth * 0.5, -10, text, "ChatFont", Color(140, 0, 0, 100), Color(255, 255, 255, 255))
     cam.End3D2D()
 
     Ang:RotateAroundAxis(Ang:Right(), 180)
 
-    cam.Start3D2D(Pos, Ang, 0.1)
+    cam.Start3D2D(self:LocalToWorld(Pos-offset), Ang, 0.1)
```
Comment (FPtje) [inline entities/entities/spawned_money/cl_init.lua]: Spaces around operators. Why haven't you installed glualint yet!?

File: entities/entities/spawned_money/cl_init.lua (hunk 1 above)
Comment (ARitz-Cracker) [inline entities/entities/spawned_money/cl_init.lua]: That is just to isolate the "z" value I get from the objects bounding box. (Which is done a few lines down)

File: entities/entities/spawned_money/cl_init.lua (hunk 2 above)
Comment (ARitz-Cracker) [inline entities/entities/spawned_money/cl_init.lua]: I use notepad++, if you really want to keep the entire DarkRP gamemode that complaint to the specs you defined in glualint, you can edit it yourself if you'd like.

File: entities/entities/spawned_money/cl_init.lua (hunk 3 above)
Comment (ARitz-Cracker) [inline entities/entities/spawned_money/cl_init.lua]: See previous comment

File: entities/entities/spawned_money/cl_init.lua (hunk 4 above)
Comment (ARitz-Cracker) [inline entities/entities/spawned_money/cl_init.lua]: Why are you so judgemental? I just wanted to help! D:

**Comment by ARitz-Cracker** [comment]
I did notice while I was coding it that it assumes the thinnest part it the "z" axis. I _could_ check each axis and determine which side is the smallest. If you'd like me to do that before implementing something like this, I'd be happy too.

**Comment by Shigbeard** [comment]
Also issues that regard the size of the label could be resolved using a config option for scaling the label.

**Comment by FPtje** [comment]
A config option would then also apply to money packets.

What's wrong with making an entity that has the `spawned_money` as base? That's the most generic and appropriate solution.

**Comment by JamesxX** [comment]
Falco, instead of being a pain to people who write code for a gamemode credited to you, simply because of their style of code, you could instead just lint it yourself. If it's as easy as you say, do it yourself.

**Comment by FPtje** [comment]
People making PRs can be reasonably expected to program in the same style as the rest of the project.

**Comment by JamesxX** [comment]
As with anything, but at the end of the day, if people are doing work for free, you can give them some allowance.

**Comment by Shigbeard** [comment]
Holy shit getting this worked up over gaps or lack thereof, jeez I have two recommendations falco.

The first is to go get a cup of coffee. Like, right this second.

The second is to calm the fuck down, it's just a few lines and doesn't even break anything.

**Comment by JamesxX** [comment]
^

File: entities/entities/spawned_money/cl_init.lua [hunk 5]
```diff
@@ -3,27 +3,25 @@ include("shared.lua")
...
         draw.WordBox(2, -TextWidth * 0.5, -10, text, "ChatFont", Color(140, 0, 0, 100), Color(255, 255, 255, 255))
```
Comment (JamesxX) [inline entities/entities/spawned_money/cl_init.lua]: Why *0.5 when you could /2. Seems more intuitive, and easier to understand.

**Comment by FPtje** [comment]
The pushy glualint stuff was meant to be a joke. Sorry if it didn't come across as such.

**Comment by FPtje** [comment]
Still though, if you want text drawn over custom models, wouldn't it be better to just create a derived entity? You'd be able to modify everything: position, angle, scale, colour, the actual text etc.

**Comment by ARitz-Cracker** [comment]
Indeed it is. I thought since there was a config option to have custom models, why not have a (basic) way to support them?

File: entities/entities/spawned_money/cl_init.lua (hunk 1 above)
Comment (FPtje) [inline entities/entities/spawned_money/cl_init.lua]: But 1 vs 0.9 though

File: entities/entities/spawned_money/cl_init.lua (hunk 2 above)
Comment (FPtje) [inline entities/entities/spawned_money/cl_init.lua]: Also, OBBMaxs - OBBCenter shouldn't get you to the top middle of an object, which is probably what you'd want. With the standard money packet, it just gets the Pos (which is either bottom middle or center middle) and adds Ang:Up() * 0.9 to it to go to the top middle of the money packet.

OBBMaxs - OBBCenter is the local vector going from the center to the top right. Surely that shouldn't go straight up for most models.

**Comment by FPtje** [comment]
That's a good point.

File: entities/entities/spawned_money/cl_init.lua (hunk 2 above)
Comment (ARitz-Cracker) [inline entities/entities/spawned_money/cl_init.lua]: It doesn't. That's why I multiply it by vector_up. So it does go straight up.

File: entities/entities/spawned_money/cl_init.lua (hunk 1 above)
Comment (ARitz-Cracker) [inline entities/entities/spawned_money/cl_init.lua]: That's why I use self:OBBMaxs() - self:OBBCenter(), to get the local pos of the "highest" position of the model. Multiplying it by 0,0,1 simply isolates the z value while setting the x and y values to 0.

**Comment by FPtje** [comment]
I tested the solution and found that it would make things look worse.

This is what the text looks like in the current master branch. Note the red slab of text being just slightly above the actual model.
hl2_2016-10-15_14-03-57


This is what the text looks like in the PR. Note that the red slab of text is much further above the model.
hl2_2016-10-15_14-02-32

Here's the cause (done with the `cl_ent_bbox` concommand):
hl2_2016-10-15_14-05-17

The bounding box of the model is bigger than the model itself. Putting the text right on top of this bounding box doesn't work. Not for money packets at least. There has to be a specific offset for money packets. That's where the `0.9` comes from in the current code. It also means that there *cannot* be an acceptable generic solution. There is no acceptable "one size fits all". It's way too different per model.

Also, I've tweaked the offset to `0.82`. That's really tight, and still looks good from some distance (with and without anti-aliasing).

**Pull request closed without merging**